framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,581 @@
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+#endif
+	return MEDIA_OK;
+}
+
+void media_shutdown(void)
+{
+	// Shutdown playback thread
+	g_playing_live = false;
+	sem_post(&g_sem_playing);
+	pthread_join(g_pth_playing, NULL);
+	pthread_mutex_destroy(&g_mutex_playing_q);
+	sem_destroy(&g_sem_playing);
+
+	// Shutdown record thread
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	g_recording_live = false;
+	sem_post(&g_sem_recording);
+	pthread_join(g_pth_recording, NULL);
+	pthread_mutex_destroy(&g_mutex_recording_q);
+	sem_destroy(&g_sem_recording);
+#endif
+}
+
+media_result_t media_play(media_t *m, bool loop)
+{
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+	return MEDIA_OK;
+}
+
+media_result_t media_record(media_t *m)
+{
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	pthread_mutex_lock(&g_mutex_recording_q);
+	dq_addlast((dq_entry_t *)m, &g_recording_q);
+	pthread_mutex_unlock(&g_mutex_recording_q);
+	sem_post(&g_sem_recording);
+#else
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+#endif
+	return MEDIA_OK;
+}
+
+media_result_t media_stop(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PLAYING || m->state == MEDIA_STATE_RECORDING) {
+		m->state = MEDIA_STATE_STOPPING;
+		return MEDIA_OK;
+	}  
+	return MEDIA_ERROR;
+}
+
+media_result_t media_pause(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PLAYING) {
+		m->state = MEDIA_STATE_PAUSING;
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+media_result_t media_resume(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PAUSED) {
+		m->state = MEDIA_STATE_PLAYING;
+		sem_post(&g_sem_playing);
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+void media_setvol(media_t *m, unsigned char vol)
+{
+	return;
+}
+
+bool media_isplaying(media_t *m)
+{
+	return (m->state == MEDIA_STATE_PLAYING);
+}
+
+media_result_t media_getvol(media_t *m)
+{
+	return MEDIA_ERROR;
+}
+
+media_result_t media_seek(media_t *m, int time_ms)
+{
+	return MEDIA_OK;
+}
+
+media_result_t media_setposition(media_t *m, int time_ms)
+{
+	return MEDIA_OK;
+}
+
+media_result_t media_getposition(media_t *m)
+{
+	return MEDIA_OK;
+}
+
+/***
+ * Functions run on threads
+ */
+
+static media_result_t read_wav_header(media_t *m)
+{
+	//struct pcm_config config;
+	struct wav_header_s header;
+	int ret;
+
+	ret = read(m->fd, &header, sizeof(header));
+	if (ret != sizeof(header)) {
+		return MEDIA_ERROR_UNKNOWN_FILE;
+	}
+
+	// Todo: Add header information to media_t
+
+	return MEDIA_OK;
+}
+
+media_t *media_open(char *path, media_op_t op, media_type_t type, media_play_type_t play_type)
+{
+	media_t *m = (media_t *)malloc(sizeof(media_t));
+	if (m == NULL) {
+		return NULL;
+	}
+
+	m->fd = -1;
+	m->op = op;
+	m->state = MEDIA_STATE_CREATED;
+	m->type = type;
+	m->play_type = play_type;
+
+	if (type == MEDIA_TYPE_WAV || type == MEDIA_TYPE_PCM) {
```

Comment:
As I remember, I already reviewed about this in last PR
I think we don't have to sperate wav & pcm .
Actually most of player check it's extension name.
enum media_type_e {
 	MEDIA_TYPE_PCM = 1,
        MEDIA_TYPE_MP3 = 2, // as a example for this review.
 	MEDIA_TYPE_UNSUPPORT = -1
};

media_t *media_open() {
     m->type = get_file_extension_type(path);
     if (type == MEDIA_TYPE_UNSUPPORT) {
             dbg("unsupport type\n");
             return NULL;
     }
     switch (m->type)
     case MEDIA_TYPE_PCM : 
            read_wave_header(); 
            break;
     case MEDIA_TYPE_MP3:
            read_mp3_header();
            break;
.
.
.

Reply:
We should differentiate between PCM and WAV. In wav case, we should read the header of the file and set the pcm parameters accordingly. In PCM case, we should fetch the parameters from the application. 
In record case, we would need to write wav header only in wav case and not in pcm case. 
In my opinion, we should keep this type.

Reply:
Ok, it is not important for now, but please conside flow of code what I mentioned above.
But make sure Mediaplayer is not low-level sound system, it is layer to support sound/music player.
Player doesn't know file's contents before read, it checks extension only.

Reply:
Ok, I will add the extension checking in media_open